ORM

tpor9095 opened the pull request:
I know this is going to be a crazy merge, but I feel now is probably the time.  All tests are passing (except for those that raise NotImplementedErrors which have been implemented by Daisy on the master branch).

I have converted everything over to using the SQLAlchemy ORM; perhaps my efforts were less than optimal, but I was trying to do it all in a relatively short period of time.

I have also tweaked a few other things here or there.  For example, comments are now accompanied by Gravatars, and each comment is stored in the database with a full Python datetime object, which is drawn out and then converted into a nice format when we need it in the views.

We need a proper code review, in my opinion, before we can reasonably move forward.

Commit:
Converted Meterage to a package structure

Commit:
Login is functional, show_entries is approaching functional

Commit:
added empty(ish) db folder

Commit:
Login, main page working except for Gravatars

I need to fix up the definition of the Entry class such that when I
obtain the user_id, I can use this to obtain the gravataremail
associated with this user_id in the users table and therefore pass
this to the function to generate the Gravatar itself.

At this stage, my lack of experience withdatabases is letting me
down.  I'm fairly sure I want to perform a join on the used_id between
entries and users tables, then somehow extract the Gravatar.  I can do
this easily with a query string, but I want this to be automatic.
When I ask for all the entries, the gravatar should come as part of
the request automatically.

Commit:
modified readme

Commit:
Wrote Comment class, began changing views.py for comments

Many comments can be associated with one Entry.  The Comment is
connected to the Entry via the entry_id field.

Commit:
Added 'admin' flag to User class

If that User is an admin then it will be True, otherwise it will be False.

Commit:
Near full functionality

The code needs a refactor and some explanatory commenting, but save
for this it is almost fully functional.

The only place that things are not fully functional is in the Gravatar
department.  I can easily get it working, but the implementation would
be somewhat of a hack.  I would prefer the Gravatart email to be a a
column in the Entry class that matches up with the User class via the
user id of the entry poster.

Commit:
removed unneeded code

Commit:
Gravatars working

Commit:
Various improvements and refactorings

Commit:
minor fix

Commit:
Incremental steps towards unit testing

Unit testing the changes is proving quite difficult.  I wil probably
have to create another separate package for testing.

Commit:
Wrote semi-functional setup and teardown methods

The set-up method sets up some config parameters in the
meterage.app.config dictionary and then reloads the meterage.models
script, which contains all of the models used in the database.  We
need to perform this reload as all of the models need to inherit the
db.Model object that we have just set up in our set up method.

At the moment we are able to pass the very first test, but there is an
issue with the next setup after the teardown.  This needs to be
resolved, but this commit represents a decent step forward; worthy of
a commit.

Commit:
Unit tests passing

All of the unit tests that were previously passing (i.e. not those
that raise a NotImplementedError) are passing now.  I am afraid that
in order ot achieve this I was forced to use somewhat of a hack; I
disabled DEBUG mode.  I think that Flask does not like SQLAlchemy
during testing, though I am not sure.

I still need to clean up the code considerably and annotate it, but at
the moment the unit tests are passing.

Commit:
Re-implemented Gravatars, halfway through behave testing

I have re-implemented Gravatars using the official Flask Gravatar
extension.  This will require a fresh `pip install -r
requirements.txt`.

I am also halfway through fixing up all of the behave testing; I am
yet to finish HashedPasswords and Comment features.

Commit:
Completed HashedPasswords.feature behave testing

Commit:
All behave tests and unit tests passing.

* Implemented filte to convert datetime objects into timestamp strings
  in the comment section.
* Implemented Gravatars on comment page for the entry being commented
  on.

Commit:
Included Gravatars with each comment

ghost commented:
Actually... I need to write tests for the ORM itself.  All the other tests are passing, but the ORM needs its own tests now. :disappointed:

Commit:
Completed 'add' test for ORM

Commit:
Finished read test for ORM

Commit:
Writing test for ORM update

Commit:
Two out of four ORM unit tests passing

There remain issues with the update and delete functions of the ORM.
It seems that when we perform an update or a delete, it is apparently
working but when we dig down into a manual SQL query on the database
the data persist when they ought not to.

Commit:
All tests passing

Commit:
ORM tests create and read passing

Commit:
ORM update test working

Note that the ORM tests need to be run on an individual basis; as part
of the test suite, they do not run correctly.

Commit:
Merge branch 'temp' into ORM

Commit:
All ORM unit tests passing

For some reason, at this point the ORM tests refuse to all pass when
running as part of the test suite.  This has something to do with
SQLAlchemy.  Perhaps something to do with the Session object, or
threads?

In any case, the tests *are* passing and this means that we are able
to proceed.

ghost commented:
I've done the merge over on the branch 'merging'
